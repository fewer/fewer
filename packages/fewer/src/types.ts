import { Association, AssociationType } from './Association';
import { Select } from '@fewer/sq';

// TODO: Move this back into unique symbols once this bug is resolved:
// https://github.com/Microsoft/TypeScript/issues/29108
export const enum INTERNAL_TYPES {
  ALL_FIELDS = '@@ALL_FIELDS',
  RESOLVED_TYPE = '@@RESOLVED_TYPE',
  INTERNAL_TYPE = '@@INTERNAL_TYPE',
  SCHEMA_TYPE = '@@SCHEMA_TYPE',
  JOINS = '@@JOINS',
  TO_SQ_SELECT = '@@TO_SQ_SELECT',
  FUNCTIONS_TYPE = '@@FUNCTIONS_TYPE',
  BUILT_TABLE = '@@BUILT_TABLE'
}

// We default to selecting all fields, but once you pluck one field, we need to remove
// the ALL_FIELDS symbol and only carry forward the plucked fields.
export type CreateSelectionSet<
  Original,
  Additional
> = Original extends INTERNAL_TYPES.ALL_FIELDS
  ? Additional
  : Original | Additional;

export interface Associations {
  [key: string]: Association;
}

type UnrollAssociation<
  T extends Association
> = WhereType<T[INTERNAL_TYPES.SCHEMA_TYPE], T[INTERNAL_TYPES.JOINS]>;

type WhereForType<T> = {
  [P in keyof T]?: NonNullable<T[P]> | NonNullable<T[P]>[]
};

type Merge<T extends object> = { [K in keyof T]: T[K] };

export type WhereType<Root, Assoc extends Associations = {}> = Merge<WhereForType<
  Root
> &
  { [P in keyof Assoc]?: UnrollAssociation<Assoc[P]> }>;

export type Subset<Root, Keys, AssociationKeys extends keyof Root> = [
  Keys
] extends [INTERNAL_TYPES.ALL_FIELDS]
  ? Root
  : { [P in ((keyof Root) & Keys) | AssociationKeys]: Root[P] };

export type ResolveAssociations<Assoc extends Associations> = {
  [P in keyof Assoc]: [Assoc[P]['type']] extends [AssociationType.HAS_MANY]
    ? Assoc[P][INTERNAL_TYPES.RESOLVED_TYPE][]
    : Assoc[P][INTERNAL_TYPES.RESOLVED_TYPE]
};

export interface CommonQuery<Type, Assoc extends Associations> {
  pluck(...fields: (keyof Type)[]): any;
  pluckAs(name: keyof Type, alias: string): any;
  where(wheres: any): any;
  limit(amount: number): any;
  offset(amount: number): any;
  order(): any;
  load(name: string, association: Association): any;
  join(name: string, association: Association): any;
  [INTERNAL_TYPES.TO_SQ_SELECT](): Select;
}
