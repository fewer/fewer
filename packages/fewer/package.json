{
  "name": "fewer",
  "version": "0.0.6",
  "description": "A minimal ORM for Node.js.",
  "keywords": [
    "orm",
    "sql",
    "fewer"
  ],
  "author": "Jordan Gensler <jgensler@netflix.com>",
  "homepage": "https://github.com/fewer/fewer",
  "license": "MIT",
  "main": "lib/index.js",
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/fewer/fewer.git"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "bugs": {
    "url": "https://github.com/fewer/fewer/issues"
  },
  "dependencies": {
    "@fewer/sq": "^0.0.4"
  },
  "devDependencies": {
    "@fewer/typeval": "^0.0.4"
  }
}
