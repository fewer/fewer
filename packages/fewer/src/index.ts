import Schema from './Schema';
import { Repository, createRepository } from './Repository';

export { Schema, Repository, createRepository };
