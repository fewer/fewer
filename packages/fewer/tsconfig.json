{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "lib",
    "composite": true,
    "rootDir": "src"
  },
  "include": ["./src"],
  "references": [
    { "path": "../sq" },
  ]
}
