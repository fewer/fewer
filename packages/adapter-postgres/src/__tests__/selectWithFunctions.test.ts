import { createSchema, createDatabase, createRepository } from 'fewer';
import { Adapter, $selectToSQL } from '../';
import { INTERNAL_TYPES } from 'packages/fewer/lib/types';

const database = createDatabase({
  adapter: new Adapter({}),
});

const VERSION = 1;

const schema = createSchema()
  .table(database, 'users', t => ({
    first_name: t.string(),
    last_name: t.string(),
  }))
  .table(database, 'posts', t => ({
    title: t.string(),
    user_id: t.bigint(),
    subtitle: t.string(),
  }));

const Users = createRepository(schema.tables.users);
const Posts = createRepository(schema.tables.posts);

describe('select with functions', () => {
  describe('eq', () => {
    it('generates good sql', () => {
      const select = Users.where((fns, c) => fns.eq(c.first_name, 'Emily'))[INTERNAL_TYPES.TO_SQ_SELECT]();
      const sql = $selectToSQL(select.context);

      expect(sql).toMatchSnapshot();
    });
  });
});