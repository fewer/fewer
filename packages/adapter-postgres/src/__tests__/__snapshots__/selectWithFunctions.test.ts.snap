// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select with functions eq generates good sql 1`] = `"SELECT * FROM users"`;
